File: exclusiveTime.py
from typing import List
from collections import deque

class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        # answer array to store the final answer
        ans = [0] * n

        # declare a stack to store the pairs
        cur = deque()

        # prev time variable
        prevTime = 0

        for s in logs:
            id = int(s[:s.find(':')])
            time = int(s[s.rfind(':') + 1:])

            if 'end' in s:
                ans[id] += time - prevTime + 1
                cur.pop()
                prevTime = time + 1
            else:
                if cur:
                    ans[cur[-1][0]] += time - prevTime
                cur.append((id, time))
                prevTime = time

        return ans

File: exclusiveTime.cpp
class Solution {
public:
     vector<int>exclusiveTime(int n, vector<string>&logs){
          //answer array to store the final answer
          vector<int>ans(n);


          //declare a stack to store the pairs
          stack<pair<int,int>>cur;


          //prev time variable 
          int prevTime = 0;

           for(auto & s: logs){

             int id = stoi(s.substr(0, s.find(':')));
             int time = stoi(s.substr(s.rfind(':') + 1));

             if(s.find('e') != -1){
                   ans[id] += time - prevTime + 1;
                   cur.pop();
                   prevTime = time + 1
             }
             else{
                  if(!cur.empty())ans[cur.top().first] += (time - prevTime);
                  cur.push({num, time});
                  prevTime = time;
             }
          }

          return ans;
     }
};
